akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/EventSourcedBehaviorTestKitImpl.scala

```diff
@@ -0,0 +1,191 @@
+      if (events.nonEmpty) events.head else throw new AssertionError("No events")
+    }
+
+    override def eventOfType[E <: Event: ClassTag]: E =
+      ofType(event, "event")
+
+    override def stateOfType[S <: State: ClassTag]: S =
+      ofType(state, "state")
+
+    override def reply: Reply = replyOption.getOrElse(throw new AssertionError("No reply"))
+
+    override def replyOfType[R <: Reply: ClassTag]: R =
+      ofType(reply, "reply")
+
+    // cast with nice error message
+    private def ofType[A: ClassTag](obj: Any, errorParam: String): A = {
+      obj match {
+        case a: A => a
+        case other =>
+          val expectedClass = implicitly[ClassTag[A]].runtimeClass
+          throw new AssertionError(
+            s"Expected $errorParam class [${expectedClass.getName}], " +
+            s"but was [${other.getClass.getName}]")
+      }
+    }
+  }
+
+  final case class RestartResultImpl[State](state: State) extends RestartResult[State]
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class EventSourcedBehaviorTestKitImpl[Command, Event, State](
+    actorTestKit: ActorTestKit,
+    behavior: Behavior[Command])
+    extends EventSourcedBehaviorTestKit[Command, Event, State] {
+
+  import EventSourcedBehaviorTestKitImpl._
+
+  private def system: ActorSystem[_] = actorTestKit.system
+
+  override val persistenceTestKit: PersistenceTestKit = PersistenceTestKit(system)
+
+  private val probe = actorTestKit.createTestProbe[Any]()
+  private val stateProbe = actorTestKit.createTestProbe[State]()
+  private var actor: ActorRef[Command] = actorTestKit.spawn(behavior)
+  private def internalActor = actor.unsafeUpcast[Any]
+  private val persistenceId: PersistenceId = {
+    internalActor ! EventSourcedBehaviorImpl.GetPersistenceId(probe.ref)
+    try {
+      probe.expectMessageType[PersistenceId]
+    } catch {
+      case NonFatal(_) =>
+        throw new IllegalArgumentException("Only EventSourcedBehavior, or nested EventSourcedBehavior allowed.")
+    }
+  }
+  private val serializationTestKit = new SerializationTestKit(system)
+
+  private var serializationSettings: SerializationSettings = SerializationSettings.enabled
+  private var emptyStateVerified = false
+
+  persistenceTestKit.clearByPersistenceId(persistenceId.id)
+
+  def setSerializationSettings(settings: SerializationSettings): EventSourcedBehaviorTestKit[Command, Event, State] = {
+    serializationSettings = settings
+    this
+  }
+
+  override def runCommand(command: Command): CommandResultImpl[Command, Event, State, Nothing] = {
+    if (serializationSettings.enabled && serializationSettings.verifyCommands)
+      verifySerializationAndThrow(command, "Command")
+
+    if (serializationSettings.enabled && !emptyStateVerified) {
+      internalActor ! EventSourcedBehaviorImpl.GetState(stateProbe.ref)
+      val emptyState = stateProbe.receiveMessage()
+      verifySerializationAndThrow(emptyState, "Empty State")
+      emptyStateVerified = true
+    }
```

Comment:
The intention was to check serialization of empty state in case it was used as snapshot. Maybe you are right that we never snapshot the empty state so that check is not strictly needed. I think I leave it in anyway.